python/cudf/cudf/core/mixins/mixin_factory.py

```diff
@@ -0,0 +1,218 @@
+        classes.
+    supported_ops : List[str]
+        The list of valid operations that subclasses of the resulting mixin may
+        request to be implemented.
+
+    Examples
+    --------
+    >>> MyFoo = _create_delegating_mixin(
+    ...     "MyFoo", "MyFoo's docstring", "FOO", "do_foo", {"foo1", "foo2"}
+    ... )
+
+    >>> # The above is equivalent to defining a class like so:
+    ... class MyFoo:
+    ...     # The set of valid foo operations.
+    ...     _VALID_FOOS = {"foo1", "foo2"}
+    ...
+    ...     # This is the method for child classes to override. Note that the
+    ...     # first parameter is always called "op".
+    ...     def do_foo(self, op, *args, **kwargs):
+    ...         raise NotImplementedError
+
+    >>> # MyFoo can be used as follows.
+    >>> class BarImplementsFoo(MyFoo):
+    ...     _SUPPORTED_FOOS = ("foo1",)
+    ...     _FOO_DOCSTRINGS = {"ret": "42"}
+    ...
+    ...     # This method's docstring will be formatted and used for all valid
+    ...     # operations. The `op` key is always automatically filled in, while
+    ...     # other keys are formatted using _FOO_DOCSTRINGS.
+    ...     def do_foo(self, op, *args, **kwargs):
+    ...         '''Perform the operation {op}, which returns {ret}.'''
+    ...         return 42
+
+    >>> bar = BarImplementsFoo()
+    >>> print(bar.foo1())
+    42
+
+    >>> # This will raise an AttributeError because foo2 is not supported by
+    >>> # the BarImplementsFoo subclass of MyFoo.
+    >>> # print(bar.foo2())
+
+    >>> # The docstring is formatted with the operation name as well as any
+    >>> # additional keys provided via the _FOO_DOCSTRINGS parameter.
+    >>> print(bar.foo1.__doc__)
+    Perform the operation foo1, which returns 42.
+    """
+    # The first two attributes may be defined on subclasses of the generated
+    # OperationMixin to indicate valid attributes and parameters to use when
+    # formatting docstrings. The supported_attr will be defined on the
+    # OperationMixin itself to indicate what operations its subclass may
+    # inherit from it.
+    validity_attr = f"_VALID_{category_name}S"
+    docstring_attr = f"{category_name}_DOCSTRINGS"
+    supported_attr = f"_SUPPORTED_{category_name}S"
+
+    class OperationMixin:
+        @classmethod
+        def __init_subclass__(cls):
+            # Only add the valid set of operations for a particular class.
+            valid_operations = set()
+            for base_cls in cls.__mro__:
+                valid_operations |= getattr(base_cls, validity_attr, set())
+
+            invalid_operations = valid_operations - supported_operations
+            assert (
+                len(invalid_operations) == 0
+            ), f"Invalid requested operations: {invalid_operations}"
+
+            base_operation = getattr(cls, category_operation_name)
+            for operation in valid_operations:
+                if operation not in dir(cls):
```

Comment:
But I doubt this will be much of an issue, because the pattern we are applying to `_base_operation` should be simple and inherited as-is in the class hierarchy (e.g. _reduce, _scan). And if certain operation needs special operations we can explicitly override:
```python
In [77]: baz = BazIsBar_AlsoImplementsFoo()

In [78]: baz.useful_foo_op()
do useful_foo_op a bit differently in foo.
```

Reply:
And finally, if two mixins have conflicting method names, the first that declares dictates. But this wouldn't be an actual issue because you can't define two different implementation of the same function today.

Reply:
This is something that I actually had support for at some point, or at least I'm pretty sure that I did. It must have gotten lost in one of the edits. I'm going to attempt to recreate what I used to do, which I think will simplify some of the logic and make it more flexible to override in subclasses.

Reply:
This turns out to be regression caused by the descriptor approach because the Operation for a class binds the base operation on construction rather than being defined in an `exec` string that calls e.g. `self._reduce`, which will always look up the base operation at call time. I've resolved it by adding some complexity to the approach for determining when an operation is overwritten. Have a look at https://github.com/rapidsai/cudf/pull/9925/commits/1303bb7a773d4ef94016762cdf35c79430ac8f3c and see if it behaves the way that you expect. I tested a number of different scenarios locally to see how inheritance and method overriding works. One note is that if a subclasses `_VALID_OPS` contains an operation that was also contained in one of its parents, it's basically a no-op which I think is what is expected but worth checking. It is a little unfortunate that the new approach requires creating new Operation objects for any subclass that overrides the base class, but I think the overhead should be limited and I think it's necessary to resolve this problem. I considered `exec` based alternatives, and they _behave_ correctly but will not patch attributes like the docstring or the function signature correctly. In order to observe all of those changes I think what I've done is necessary.

Reply:
Thanks. Current commit does what's expected above. When user explicitly defines `base_operation` in subclass, all operations will be redefined with the new `base_operation`.

Reply:
Thanks. Current commit does what's expected. To sum up, when user explicitly redefines `base_operation` in the subclass, all operations will get overridden too.